### ISSUE TITLE
Make reproducing test(s) for failing to generalize program clauses

### ISSUE THREAD

**Issue by nathanwhit** [rust-lang/chalk#614]
If program clauses are not properly generalized it can cause issues in the recursive solver. This was the cause of the recent [breakage in rust-analyzer](https://github.com/rust-analyzer/rust-analyzer/issues/6078). We should have a test that covers this to prevent future issues.

[Discussed briefly here](https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/Recent.20breakage.20in.20recursive.20solver).

**Comment by detrumi** [comment]
Minimal example:
```
fn main() {
    foo(|a: u32| {});
}

fn foo<F>(f: F)
where
    F: Send,
{
}
```

Trying to fit this in a test is tricky though, as the problem triggers with this goal:
```
Implemented({closure:ClosureId(0)}<for<0> Safe () [?0 := Uint(U32), ?1 := ^1.0]>: Send)
```
The syntax has changed slightly since, but this is a `TyKind::Closure` with the substitution being a `FnPointer` (`for<0> Safe () [?0 := Uint(U32), ?1 := ^1.0]`).

I don't think there's currently a way to trigger closure substitutions in chalk tests. My attempts so far:
```
test! {
        program {
            #[lang(fn_once)]
            trait FnOnce<Args> {
                type Output;
            }

            #[auto] trait Send { }

            fn foo<F>(f: F) -> ()
                where F: Send;

            closure bar(self, a: u32) {}
        }

        goal {
            <bar as FnOnce<(foo<(bar, u32)>)>>: Send // Not valid test syntax
        } yields {
            "Unique"
        }
    }
```

**Comment by flodiebold** [comment]
Yes, I've had similar problems reproducing problems with closures from RA. Maybe the closure support in the test harness needs to be more flexible somehow. I've usually been able to reproduce the problems by just modeling the closure as a normal struct.
